libwild/src/linker_script.rs

```diff
@@ -33,11 +34,36 @@ pub(crate) fn linker_script_to_inputs(
         .into_iter()
         .map(|mut input| {
             input.search_first = Some(directory.to_owned());
+            if let (Some(sysroot), InputSpec::File(file)) = (sysroot, &mut input.spec) {
+                if let Some(new_file) = maybe_apply_sysroot(path, file, sysroot) {
+                    *file = new_file;
+                }
+            }
+
             input
         })
         .collect())
 }
 
+fn maybe_apply_sysroot(
+    linker_script_path: &Path,
+    input_path: &Path,
+    sysroot: &Path,
+) -> Option<Box<Path>> {
+    if matches!(linker_script_path.canonicalize(), Ok(p) if p.starts_with(sysroot)) {
```

Comment:
`linker_script_path` can look like this: `/usr/lib/gcc/aarch64-linux-gnu/14.2.0/../../../../aarch64-linux-gnu/lib/../lib64/libgcc_s.so` and because `Path::starts_with` checks the path literally, without `Path::canonicalize` it would give false negative if we compare against `/usr/lib/aarch64-linux-gnu`.

`canonicalize` is the easy way to avoid carrying about it but not necessarily the best: it'll call libc to check if file exists and normalize (so, allocate) on every iteration.

Reply:
Okay, I'm probably trying harder than necessary. Without `--sysroot` this won't affect performance at, with `--sysroot` hopefully it will be acceptable.

Reply:
If we're worried, there's always https://docs.rs/normalize-path/0.2.1/normalize_path/ as a possibility. Or if we worry about too many dependencies, we could roll our own, since it presumably would be a pretty simple function to implement.

Doing normalisation or canonicalisation earlier so we only do it once might also help, although I guess we wouldn't want to incur that cost if we didn't have a sysroot set.

But you're right, it's probably fine as-is.

Reply:
I had something similar to that, but without creating intermediate PathBuf. However it looked unnecessarily complex for a problem that maybe isn't really a problem. Although I can use `normalize_path` if you want, an allocation in the loop is still better than allocation with IO.

Reply:
Went for it, even readability has improved.